typed-racket-lib/typed-racket/typecheck/tc-lambda-unit.rkt

```diff
@@ -271,146 +296,222 @@
+  (for*/list ([arrow (in-list arrows)]
+              [dom (in-value (Arrow-dom arrow))]
+              [rst (in-value (Arrow-rst arrow))]
+              #:unless (in-arities? seen arrow)
+              #:when (cond
+                       [formals-rest?
+                        (or (Type? rst) (>= (length dom) pos-count))]
+                       [rst (<= (length dom) pos-count)]
+                       [else (= (length dom) pos-count)]))
+    arrow))
+
+
+
+
+;; formals+bodies  : formals and bodies to check
+;; expected-arrows : expected arrow types for the overall case-lambda
+;; orig-arrows     : an association list recording if any formals and bodies
+;;                   have _already_ been checked at a certain Arrow type
+(define/cond-contract
+  (check-mono-lambda/expected formals+bodies expected-arrows orig-arrows)
+  (-> (listof (cons/c formals? syntax?))
+      (listof Arrow?)
+      (listof (cons/c (cons/c formals? syntax?)
+                      (listof Arrow?)))
+      (listof Arrow?))
+
+  ;; arities we have seen so far while checking case-lambda clauses
+  (define seen initial-case-arities)
+  ;; For each clause we figure out which arrows it needs to type check at,
+  ;; and also which clauses are dead code.
+  (define to-check-list
+    (for*/list ([f+b (in-list formals+bodies)]
+                [clause-formals (in-value (car f+b))]
+                [clause-body (in-value (cdr f+b))])
+      (define matching-arrows
+        (arrows-matching-seen+formals expected-arrows
+                                      seen
+                                      clause-formals))
+      (when (or (in-arities? seen clause-formals)
+                (null? matching-arrows))
+        (warn-unreachable clause-body)
+        (add-dead-lambda-branch (formals-syntax clause-formals)))
+      (set! seen (add-to-arities seen clause-formals))
+      (cons f+b matching-arrows)))
 
-;; Currently no support for objects representing the rest argument
-(define (formals->objects formals)
-  (for/list ([i (in-list (formals-positional formals))])
-    (make-Path null i)))
+  (cond
+    [(and (andmap (λ (f+b+arrows) (null? (cdr f+b+arrows)))
+                  to-check-list)
+          ;; If the empty function is expected, then don't error out
+          (not (null? expected-arrows)))
+     ;; TODO improve error message.
+     (tc-error/expr #:return (list (-Arrow null -Bottom #:rest Univ))
+                    "Expected a function of type ~a, but got a function with the wrong arity"
+                    (make-Fun expected-arrows))]
+    [else
+     (apply
+      append
+      (for*/list ([(f+b arrows) (in-assoc to-check-list)]
+                  [clause-formals (in-value (car f+b))]
+                  [clause-body (in-value (cdr f+b))]
+                  [orig-arrows (in-value (cond
+                                           [(assoc f+b orig-arrows) => cdr]
+                                           [else '()]))]
+                  [arrow (in-list arrows)])
```

Comment:
Note that checking each one separately is less complete than intersecting everything in `arrows`, but is sound. Maybe put a comment here to that effect.